src/agent_observer.py

```diff
@@ -0,0 +1,104 @@
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+        self.jenkins_agent_service = jenkins_agent_service
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_joined, self._on_agent_relation_joined
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_changed, self._on_agent_relation_changed
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
+        )
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired when an agent has joined the relation.
+        """
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
+
+        relation_data = self.state.agent_meta.get_jenkins_agent_v0_interface_dict()
+        logger.debug("Agent relation data set: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
+
+    def _on_agent_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Handle agent relation changed event.
+
+        Args:
+            event: The event fired when the agent relation data has changed.
+        """
+        logger.info("%s relation changed.", event.relation.name)
+
+        # Check if the pebble service has started and set agent ready.
+        if os.path.exists(str(AGENT_READY_PATH)) and self.jenkins_agent_service.is_active:
+            logger.warning("Given agent already registered. Skipping.")
+            return
+
+        if not self.state.agent_relation_credentials:
+            self.charm.unit.status = ops.WaitingStatus("Waiting for complete relation data.")
+            logger.info("Waiting for complete relation data.")
+            return
+
+        self.charm.unit.status = ops.MaintenanceStatus("Starting jenkins agent service")
+        self.jenkins_agent_service.restart()
+
+        if not self.jenkins_agent_service.is_active:
+            # The jenkins server sets credentials one by one, hence if the current credentials are
+            # not for this particular agent, the agent operator should wait until it receives one
+            # designated for it.
+            logger.warning(
+                "Failed credential for agent %s, will wait for next credentials to be set",
+                self.state.agent_meta.name,
+            )
+            self.charm.unit.status = ops.WaitingStatus(
+                "Failed to start the service. Waiting for another set of credentials"
+            )
+            return
+
+        self.charm.unit.status = ops.ActiveStatus()
+
+    def _on_agent_relation_departed(self, _: ops.RelationDepartedEvent) -> None:
+        """Handle agent relation departed event."""
+        self.jenkins_agent_service.stop()
```

Comment:
Should there be any error handling here?

Reply:
The way I implemented the `stop` method makes it ignore exceptions and outputs a log message when it catches one since I think exceptions on stopping the service should not block or put the charm into error state. What do you think?

Reply:
It depends on whether the exception means that stopping the charm failed, if it did error state might be worthwhile or blocked if the suer can do something to resolve the issue

Reply:
I added error handling to set the charm to blocked with the relevant message if an exception occurs.